### PR TITLE
[feat]: [PL-24120]: Added functor for variables

### PR DESCRIPTION
**Pull request by nishantsaini20**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/32676)

### CODE REVIEW

**Review by piyush-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32676)* status: 9 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17eHrq7W6UWZnkZlRm:-N17eHrr37CHDRWuqnTK:bdgsyu) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L63)):*
> ```Java
>         variableDTOList = variableService.list(ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), null);
>         break;
>       case PROJECT:
> ```

Why an empty case for Project Scope? Default should throw error in case Scope is wrong.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 84 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17e_a36UjEXB9ynU_p:-N17e_a42mC3UWVEFkhf:buo9nmt) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L84)):*
> ```Java
>             variableService.get(ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), null, identifier);
>         break;
>       case PROJECT:
> ```

Same here
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 89 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ecUt5BGj7zxvYWcl:-N17ecUt5BGj7zxvYWcm:b-pt0y3r) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L89)):*
> ```Java
>             ngAccess.getAccountIdentifier(), ngAccess.getOrgIdentifier(), ngAccess.getProjectIdentifier(), identifier);
>     }
>     if (variableDTO.isPresent()) {
> ```

Could this be null?
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ee_u7iQYg5t7tyzw:-N17ee_u7iQYg5t7tyzx:bvrt2yj) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L90)):*
> ```Java
>     }
>     if (variableDTO.isPresent()) {
>       return variableDTO.get().getVariable().getVariableConfig().getValue();
> ```

What if anything is. Null here

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32676)* status: 9 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 63 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17eHrq7W6UWZnkZlRm:-N185Y8NAst84H4ou26j:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L63)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Why an empty case for Project Scope? Default should throw error in case Scope is wrong.
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 84 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17e_a36UjEXB9ynU_p:-N185XTq-j9ew4BpFiGT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L84)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Same here
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 89 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ecUt5BGj7zxvYWcl:-N17ll760pUVodIBNFzK:b-fvgf4t) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L89)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

Could this be null?
</blockquote></details>

No this can never be null because the get method can return empty optional but never null.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ee_u7iQYg5t7tyzw:-N17m9UNAnvLtMXaxPfH:bb9pauk) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L90)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

What if anything is. Null here
</blockquote></details>

The create method takes care that variable config is never null. So ideally anything here null means either incomplete data in DB or mapper from entity to DTO is buggy. In any case null pointer exception is expected to be thrown.

**Comment by piyush-harness** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ee_u7iQYg5t7tyzw:-N1CU2dw2sW5sHJy6STi:bqff9ns) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L90)):*
<details><summary><i>Previously, nishantsaini20 (Nishant Saini) wrote…</i></summary><blockquote>

The create method takes care that variable config is never null. So ideally anything here null means either incomplete data in DB or mapper from entity to DTO is buggy. In any case null pointer exception is expected to be thrown.
</blockquote></details>

NullPointer as response is not a graceful fallback, We should throw specific error incase something which was expected to be mandatory and is not set.

File: 120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.ws.rs.NotFoundException;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class VariableFunctor implements SdkFunctor {
+  public static final String VARIABLE = "variable";
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java:33]: Should we move this to NGCommonCostants?

File: 120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+
+  private ScopeLevel getScope(String scope) {
+    switch (scope) {
+      case "account":
+        return ScopeLevel.ACCOUNT;
+      case "org":
+        return ScopeLevel.ORGANIZATION;
+      case "project":
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java:116]: We must already have some constants for these. If not can we define?

File: 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+  public void setup() {
+    MockitoAnnotations.initMocks(this);
+  }
+
+  @Test
+  @Owner(developers = NISHANT)
+  @Category(UnitTests.class)
+  @PrepareForTest(AmbianceUtils.class)
+  public void testGet_getVariableValueMapAccount() {
+    //<+variable.account.var1>
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java:78]: remove

File: 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    expectedMap.put(variableIdentifier, fixedValue);
+    assertThat(variableMap).isEqualTo(expectedMap);
+  }
+
+  @Test
+  @Owner(developers = NISHANT)
+  @Category(UnitTests.class)
+  @PrepareForTest(AmbianceUtils.class)
+  public void testGet_getVariableValueMapOrg() {
+    //<+variable.org.var1>
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java:104]: same

File: 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    expectedMap.put(variableIdentifier, fixedValue);
+    assertThat(variableMap).isEqualTo(expectedMap);
+  }
+
+  @Test
+  @Owner(developers = NISHANT)
+  @Category(UnitTests.class)
+  @PrepareForTest(AmbianceUtils.class)
+  public void testGet_getVariableValueMapProject() {
+    //<+variable.project.var1>
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java:130]: same

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32676)* status: 9 file(s) to review. 1 change requests. 9. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 90 at r1](https://reviewable.io/reviews/harness/harness-core/32676#-N17ee_u7iQYg5t7tyzw:-N1DDXGNBmC5VbZ6Yn34:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/7c3984d8c15782f9750c2c914633cd15971d3767/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L90)):*
<details><summary><i>Previously, piyush-harness (Piyush Patel) wrote…</i></summary><blockquote>

NullPointer as response is not a graceful fallback, We should throw specific error incase something which was expected to be mandatory and is not set.
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/32676#-N0phqT_Jw7Q_yJJ2w1F-r2-33:-N1D-4AlDP3yqu94nXw-:bargbbw) ([raw file](https://github.com/harness/harness-core/blob/0b7ba934a86eab8737bd9d47a998e94343bb287d/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L33)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

Should we move this to NGCommonCostants?
</blockquote></details>

This is functor specific, so we can leave it here.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java` line 116 at r2](https://reviewable.io/reviews/harness/harness-core/32676#-N0phqT_Jw7Q_yJJ2w1F-r2-116:-N1DGuBfEO7cXV_ATe9w:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0b7ba934a86eab8737bd9d47a998e94343bb287d/120-ng-manager/src/main/java/io/harness/ng/core/variable/expressions/functors/VariableFunctor.java#L116)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

We must already have some constants for these. If not can we define?
</blockquote></details>

Done.
___
*[`120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java` line 78 at r2](https://reviewable.io/reviews/harness/harness-core/32676#-N0phqT_Jw7Q_yJJ2w1H-r2-78:-N1DGvfT6oRDk1DbLGIy:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0b7ba934a86eab8737bd9d47a998e94343bb287d/120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java#L78)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

remove
</blockquote></details>

Done.
___
*[`120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java` line 104 at r2](https://reviewable.io/reviews/harness/harness-core/32676#-N0phqT_Jw7Q_yJJ2w1H-r2-104:-N1DGwOo-8wBW3g_rPWs:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0b7ba934a86eab8737bd9d47a998e94343bb287d/120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java#L104)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

same
</blockquote></details>

Done.
___
*[`120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java` line 130 at r2](https://reviewable.io/reviews/harness/harness-core/32676#-N0phqT_Jw7Q_yJJ2w1H-r2-130:-N1DGwmx9NrRbPqw4Zdv:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/0b7ba934a86eab8737bd9d47a998e94343bb287d/120-ng-manager/src/test/java/io/harness/ng/core/variable/expressions/functors/VariableFunctorTest.java#L130)):*
<details><summary><i>Previously, meenakshi-harness (Meenakshi Raikwar) wrote…</i></summary><blockquote>

remove  all commented code
</blockquote></details>

Done.

**Review by piyush-harness — APPROVED** [review]
> Reviewed 7 of 9 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32676)* status: 9 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
:lgtm:

> Reviewed 7 of 9 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/32676#-:-N1DJ-hZ8s4MoNSPGV48:bnfp4nl)* status: :shipit: complete! ✓

**Pull request merged**